stac-utils/pgstac
# Add dockerized development environment; pypgstac testing; CI - fix issue with ingest

Related Issues: #6 

This PR adds a dockerized development environment that allows consistent testing across dev environments.

This follows the "[Scripts To Rule Them All](https://github.com/github/scripts-to-rule-them-all)" pattern for scripts that control the development environment. It makes it very easy for developers to clone the repository and get started working with the project.

A docker-compose defines the development database (which uses the existing Dockerfile in the repo) and a dev container (which uses a new Docker.dev), and all scripts run through those docker containers.

This also adds black as a formatter, mypy as a type checker, and flake8 as a linter to the python project. This set of dev tools are become standard across python projects in stac-utils.

The following modifications were made to pypgstac:
- Added the "pgready" command. This command has been useful in downsteram projects and I think it's a nice fit to include in pypgstac directly. This is used when brining up the development environment to ensure the database is ready to accept connections before doing other things, like running migrations.
- Adds type information, and refactors some of the code to make mypy happy. These type annotations and refactors shouldn't change behavior.
- Fixes an issue with loading ndjson of items. The fix was discovered by @bitner. I implemented a test to ensure test data loading works, which was failing prior to that fix.

There is no database schema updates, but I did change `pgstac.sql` to include running the `sql/999_version.sql` against new databases. Otherwise there's no entry in the migration table, and so a `pypgstac migrate` command will assume it's a new database, and fail on creation of tables. I also fixed that sql to include pypgstac in the SEARCH_PATH.

GitHub Workflows are created to perform CI on pull requests and on the main branch. It will also publish pypgstac to pypi on tags. This will allow us to publish only via tagging releases.

I was unable to get `test/test.sh` working - it seems like there's some missing files (e.g. test/test.sql) that perhaps were not committed.

I didn't touch the Makefile, but I suggest after this PR goes in to remove that file. The only thing not covered by these STRTA/CI is the publishing to dockerhub for pgstac images, which should be added in a follow up PR.

Testing:
- Clone a fresh repo
- Run `scripts/setup` to build a new docker dev environment
- Run `scripts/test` to run all the tests
- Run `scripts/console` to drop into a dev console, which will allow you to run `pypgstac` against the development db from the command line
- Run `scripts/console --db` to drop into a `psql` interactive session to inspect the database
- Run `scripts/format` to format the python code
- Run `scripts/migrate` to migrate the db (which should be a no-op)
- Run `scripts/update` to rebuild the docker containers.
